File: server/scripts/create-preloaded-dump.js
require('dotenv').config();
const { execFile } = require('child_process');
const path = require('path');
const fs = require('fs');

const OUTPUT = path.join(__dirname, '..', '..', 'db', 'preloaded.sql');

function parseDatabaseUrl(url) {
  const parsed = new URL(url);
  return {
    host: parsed.hostname,
    port: parsed.port || '5432',
    user: decodeURIComponent(parsed.username),
    password: decodeURIComponent(parsed.password),
    database: parsed.pathname.replace(/^\//, ''),
  };
}

function runPgDump(db) {
  return new Promise((resolve, reject) => {
    const args = [
      '--host', db.host,
      '--port', db.port,
      '--username', db.user,
      '--dbname', db.database,
      '--data-only',
      '--table', 'applications',
      '--table', 'frns',
      '--table', 'status_history',
      '--table', 'usac_sync_state',
      '--file', OUTPUT,
    ];

    const child = execFile('pg_dump', args, {
      env: { ...process.env, PGPASSWORD: db.password },
    });

    child.on('error', reject);
    child.on('exit', (code) => {
      if (code === 0) resolve();
      else reject(new Error(`pg_dump failed with exit code ${code}`));
    });
  });
}

function runDockerPgDump() {
  const container = process.env.PG_DUMP_CONTAINER || 'erate-471-postgres';
  const tmpPath = '/tmp/preloaded.sql';

  return new Promise((resolve, reject) => {
    const dumpArgs = [
      'exec', container,
      'pg_dump', '-U', 'erate', '-d', 'erate_471',
      '--data-only',
      '--table', 'applications',
      '--table', 'frns',
      '--table', 'status_history',
      '--table', 'usac_sync_state',
      '-f', tmpPath,
    ];

    const dump = execFile('docker', dumpArgs, (dumpErr) => {
      if (dumpErr) return reject(dumpErr);

      const copy = execFile('docker', ['cp', `${container}:${tmpPath}`, OUTPUT], (copyErr) => {
        if (copyErr) return reject(copyErr);
        resolve();
      });
      copy.on('error', reject);
    });
    dump.on('error', reject);
  });
}

async function main() {
  const databaseUrl = process.env.DATABASE_URL;
  if (!databaseUrl) {
    throw new Error('DATABASE_URL is required');
  }

  try {
    const db = parseDatabaseUrl(databaseUrl);
    await runPgDump(db);
  } catch (err) {
    if (err.code !== 'ENOENT') throw err;
    console.log('pg_dump not found locally — using Docker container dump…');
    await runDockerPgDump();
  }

  const sizeMb = (fs.statSync(OUTPUT).size / (1024 * 1024)).toFixed(1);
  console.log(`Wrote ${OUTPUT} (${sizeMb} MB)`);
  console.log('Rebuild the Docker image to bake this dump into the container.');
}

main().catch((err) => {
  console.error('Failed to create preloaded dump:', err.message);
  console.error('Requires pg_dump in PATH and an already-imported database.');
  process.exit(1);
});

File: backend/create_preloaded_dump.py
"""Create db/preloaded.sql from the current PostgreSQL database."""

import os
import subprocess
import sys
from pathlib import Path
from urllib.parse import urlparse

OUTPUT = Path(os.getenv("PRELOADED_DUMP_PATH", "/app/db/preloaded.sql"))
TABLES = [
    "tracker_applications",
    "tracker_frns",
    "tracker_status_history",
    "usac_sync_state",
]


def main() -> None:
    url = os.getenv("DATABASE_URL")
    if not url:
        print("DATABASE_URL required", file=sys.stderr)
        sys.exit(1)

    parsed = urlparse(url)
    OUTPUT.parent.mkdir(parents=True, exist_ok=True)
    env = {**os.environ, "PGPASSWORD": parsed.password or ""}

    args = [
        "pg_dump",
        "-h", parsed.hostname or "localhost",
        "-p", str(parsed.port or 5432),
        "-U", parsed.username or "",
        "-d", parsed.path.lstrip("/"),
        "--data-only",
        "--column-inserts",
        *sum([["-t", t] for t in TABLES], []),
        "-f", str(OUTPUT),
    ]
    result = subprocess.run(args, env=env, capture_output=True, text=True)
    if result.returncode != 0:
        print(result.stderr, file=sys.stderr)
        sys.exit(1)

    size_mb = OUTPUT.stat().st_size / (1024 * 1024)
    print(f"Wrote {OUTPUT} ({size_mb:.1f} MB)")


if __name__ == "__main__":
    main()